source/diffHandler.py

```diff
@@ -0,0 +1,192 @@
+from typing import List
+
+
+class DiffAlgo(AutoPropertyObject):
+	@abstractmethod
+	def diff(self, newText: str, oldText: str) -> List[str]:
+		raise NotImplementedError
+
+	@abstractmethod
+	def _getText(self, ti: TextInfo) -> str:
+		raise NotImplementedError
+
+
+class DiffMatchPatch(DiffAlgo):
+	"""A character-based diffing approach, using the Google Diff Match Patch
+	library in a proxy process (to work around a licence conflict).
+	"""
+	#: A subprocess.Popen object for the nvda_dmp process.
+	_proc = None
+	#: A lock to control access to the nvda_dmp process.
+	#: Control access to avoid synchronization problems if multiple threads
+	#: attempt to use nvda_dmp at the same time.
+	_lock = Lock()
+
+	def _initialize(self):
+		"Start the nvda_dmp process if it is not already running."
+		with DiffMatchPatch._lock:
+			if not DiffMatchPatch._proc:
+				log.debug("Starting diff-match-patch proxy")
+				if hasattr(sys, "frozen"):
+					dmp_path = (os.path.join(globalVars.appDir, "nvda_dmp.exe"),)
+				else:
+					dmp_path = (sys.executable, os.path.join(
+						globalVars.appDir, "..", "include", "nvda_dmp", "nvda_dmp.py"
+					))
+				DiffMatchPatch._proc = subprocess.Popen(
+					dmp_path,
+					creationflags=subprocess.CREATE_NO_WINDOW,
+					bufsize=0,
+					stdin=subprocess.PIPE,
+					stdout=subprocess.PIPE
+				)
+
+	def _getText(self, ti: TextInfo) -> str:
+		return ti.text
+
+	def diff(self, newText: str, oldText: str) -> List[str]:
+		try:
+			if not newText and not oldText:
+				# Return an empty list here to avoid exiting
+				# nvda_dmp uses two zero-length texts as a sentinal value
+				return []
+			self._initialize()
+			with DiffMatchPatch._lock:
+				old = oldText.encode("utf-8")
+				new = newText.encode("utf-8")
+				# Sizes are packed as 32-bit ints in native byte order.
+				# Since nvda and nvda_dmp are running on the same Python
+				# platform/version, this is okay.
+				tl = struct.pack("=II", len(old), len(new))
+				DiffMatchPatch._proc.stdin.write(tl)
+				DiffMatchPatch._proc.stdin.write(old)
+				DiffMatchPatch._proc.stdin.write(new)
+				buf = b""
+				sizeb = b""
+				SIZELEN = 4
+				while len(sizeb) < SIZELEN:
+					try:
+						sizeb += DiffMatchPatch._proc.stdout.read(SIZELEN - len(sizeb))
+					except TypeError:
+						pass
+				(size,) = struct.unpack("=I", sizeb)
+				while len(buf) < size:
+					buf += DiffMatchPatch._proc.stdout.read(size - len(buf))
+				return [
+					line
+					for line in buf.decode("utf-8").splitlines()
+					if line and not line.isspace()
+				]
+		except Exception:
+			log.exception("Exception in DMP, falling back to difflib")
+			return Difflib().diff(newText, oldText)
+
+	def _terminate(self):
+		with DiffMatchPatch._lock:
+			if DiffMatchPatch._proc:
+				log.debug("Terminating diff-match-patch proxy")
+				# nvda_dmp exits when it receives two zero-length texts.
+				DiffMatchPatch._proc.stdin.write(struct.pack("=II", 0, 0))
+				DiffMatchPatch._proc = None

```

Comment:
A question remains about whether this should wait for the process to exit. I think it probably should.